### ISSUE TITLE
It looks like Neotree has been configured for worse discoverability.

### ISSUE THREAD

**Issue by BrianMcKeever** [syl20bnr/spacemacs#6867]
I'm bran new to emacs, so I had my start directory configured to the emacs installation folder which means I'd have to navigate out of that directory to open anything on emacs. I didn't know how to do that.

I tried various hotkeys that would make sense to me like shift-tab, alt-up arrow (the win10 hotkey), pressing enter  and h on the "/users/yxven/whatever" line. Nothing worked.

I pressed space to see if there was anything there. I didn't see anything related to neotree. (Apparently, I should have looked harder here. I know SPC h k would have helped now. Whatever, I'm new.)

I start googling. I find out the plugin I'm dealing with is neotree. I check the neotree docs. The neotree docs mention nothing about going to to a previous directory. https://www.emacswiki.org/emacs/NeoTree

Although, the screenshot suggests this isn't a problem for the default neotree installation because the default neotree installation has a line that says ".. (up a dir)" at the very top. It also has a "press ? for neotree help." Both of these are disabled in spacemacs.

In any case, I googled around a lot more and eventually learned that I should type K. It took me around an hour to learn how to navigate to the previous directory. 

tldr:
neo-show-updir-line should be enabled by default. Disabling it hurts discoverability for no good reason.

**Comment by bmag** [comment]
First, I disagree with this part of your tldr:
> Disabling it hurts discoverability

It doesn't hurt discoverability. Having `..` doesn't teach you anything about `K` (or `U`). It does hurt functionality for new users.

Second, there are two issues here: discoverability of NeoTree, and the value of `neo-show-updir-line`. About `neo-show-updir`:

> neo-show-updir-line should be enabled by default

Yes and no. It's helpful for new users, but for users who are familiar with `K` (or `U`) it's a wasted line on the screen. Another argument against is that `K` (or `U`) is faster and easier to use, so it's The Right Way (TM) to go up a directory, and enabling `..` discourages correct usage because it will prevent the user from learning about `K` (or `U`). Personally, I don't have a strong opinion on the matter, either choice is fine by me.

Regarding discoverability, Spacemacs has plenty of discoverability  options, some of which you have mentioned. However, you totally missed [NeoTree in Spacemacs docs](http://spacemacs.org/doc/DOCUMENTATION.html#orgheadline120), also available in [the develop docs](http://develop.spacemacs.org/doc/DOCUMENTATION.html#neotree-file-tree) (warning: not always up-to-date with develop branch) and from within Spacemacs as the first entry of `SPC h SPC` (and then search for "neotree"). It is important to note that in unconfigured NeoTree, `h` and `?` are bound to `describe-mode`. We re-bind the `h` on purpose, but re-binding the `?` is due to evilification. We should explicitly bind it to `describe-mode` so the binding isn't lost. It's probably more useful than `evil-search-backward` (which has its uses) in this buffer.

We could do these two things to improve NeoTree in Spacemacs, regardless of the `neo-show-updir` decision: (PR welcome)
- bind `?` to `describe-mode` in NeoTree (it should be obvious where to add it when reading the NeoTree's configuration in the source)
- Some of key descriptions in our docs for NeoTree are laconic and can be improved (e.g. add verbs like "go to" or "open")

**Comment by Yxven** [comment]
Functionality for new users IS discoverability. Having to go to the docs is when discoverability has completely failed.

I don't think having the .. line is that big of a deal. That said, it only solves discovering opening the parent directory. It doesn't help discovering any other features of neotree. 

How difficult would it be to readd the "Press ? for neotree help" except instead of "?" have "SPC h d m"? Rebinding "?" doesn't seem ideal if it'll only bring up help in this particular window on spacemacs. I like the idea of writing better descriptions. I'm not sure I would have read "neotree-select-up-node" and realized that was the option I was looking for.

**Comment by bmag** [comment]
Maybe we can use `neo-banner-message`. I'm on mobile so can't check. "?" is used in other places, is intuitive, and works also in emacs editing style, so I think we should stick with it.

**Comment by nixmaniack** [comment]
I agree with @bmag. Once you use Spacemacs for a while you'll discover `?` is used for finding help. 

e.g. `SPC ?` for finding keybindings, `?` in ediff mode to get help popup, `?` in every magit buffer gives you help, neotree's default binding as well is `?` which essentially executes `describe-mode` but currently it's shadowed which we can fix.

> How difficult would it be to readd the "Press ? for neotree help" except instead of "?" 

Would be as simple as `(setq neo-banner-message "Press ? for neotree help")` which can be added [here](https://github.com/syl20bnr/spacemacs/blob/d94d0d9f7186893eb0369dbba5fd29f9a51d0397/layers/+spacemacs/spacemacs-ui-visual/packages.el#L167) but only after fixing the shadowed binding.

**Comment by deb0ch** [comment]
I just tried out neotree for the first time and it is true that going up a directory is really not obvious, it *is* a lack of discoverability.

It would have been a lot of trouble if I hadn't read about `K` here.

While `neo-show-updir-line` is maybe not the most convenient way to go and wastes a line, I agree with @Yxven that it is still a good help for the newcomer: it is obvious.

I also like the idea of binding `?` to some help, but:

- it is needed to display somewhere that `?` will show some help, as it is again not obvious
- it would waste a line anyways
- as @Yxven said, it can also be not obvious that `neo-select-up-node` will go up a directory, so something more readable than `describe-mode` would be needed.  
Maybe a basic usage docstring in a popup ?
- `U` is not described in the docs
- if you end up in insert state, `K` doesn't work anymore (but `U` still does)

**Comment by bmag** [comment]
I made a PR (#6910) to use `?` for help and improved the docs a bit. I didn't add `..` because I think `?` should be enough to discover `K` and other bindings.

@deb0ch I used `describe-mode` for `?` because is the only ready-made option I'm aware of. You are welcome to suggest a concrete alternative. By the way, `describe-mode` acts as a popup and shows NeoTree's docstring, so maybe you can resolve the issue by improving the docstring of `neotree-mode` upstream. As for `K` not working in insert state, due to evilification it shouldn't be possible to enter insert state in NeoTree, unless you use emacs editing style, press `C-z` (toggles emacs state) or change the state explicitly with `M-x evil-<state>-state`.

**Comment by nixmaniack** [comment]
We can probably create a transient state(hydra) for just discoverability I think. It won't serve much purpose apart from getting hints when needed. e.g. We have layout TS where if you press `?` you get detailed help doc of keybindings. Something similar might help. Just a thought.

**Comment by bmag** [comment]
@nixmaniack sounds like a good idea, make the TS persistent and `?` to toggle it. Using TS for key hints reminded me of [discover.el](https://www.masteringemacs.org/article/discoverel-discover-emacs-context-menus).

**Comment by Andre0991** [comment]
I created a transient state for mu4e a while ago, but I removed it as I realizes that the keybindings were really simple and using a transient state wasn't worth it.

On second thoughts, it would be nice for discoverability, just like this case.

**Comment by nixmaniack** [comment]
@bmag Yes, whatever we've added in the doc can go in TS and `?` to toggle it is perfect. 

In future, from keybindings discoverability point of view, spacemacs could go for convention such as `SPC m ?` or `M-SPC` to get `mode` specific TS - for help doc of keybindings. Although I am not sure if it'd be useful given `which-key`'s presence all over unless we _curate_ TS help doc carefully. `org-agenda` TS serves similar purpose which is bound to `M-SPC` in agenda buffer currently.

**Comment by bmag** [comment]
I tried to implement a TS just now, but I'm trouble with the formatting of the TS's doc. I thought there'd be some automatic alignment of the keys if I'd just use the following, but apparently there isn't:
```elisp
(spacemacs|define-transient-state neotree
  :title "NeoTree Key Hints"
  :bindings
  ("TAB" 'neotree-stretch-toggle "shrink/enlarge window")
  ("RET" 'neotree-enter "open file/directory")
  ("|" 'neotree-enter-vertical-split "open in vertical split")
  ("-" 'neotree-enter-horizontal-split "open in horizontal split")
  ("?" nil "close help" :exit t)
  ("c" 'neotree-create-node "create new file/directory")
  ("d" 'neotree-delete-node "delete node")
  ("gr" 'neotree-refresh "refresh")
  ("h" 'spacemacs/neotree-collapse-or-up "move up or collapse directory")
  ("H" 'neotree-select-previous-sibling-node "previous sibling")
  ("J" 'neotree-select-down-node "move down")
  ("K" 'neotree-select-up-node "move up")
  ("l" 'spacemacs/neotree-expand-or-open "open file or expand directory")
  ("L" 'neotree-select-next-sibling-node "next sibling")
  ("r" 'neotree-rename-node "rename file/dirctory")
  ("R" 'neotree-change-root "change NeoTree root")
  ("s" 'neotree-hidden-file-toggle "toggle hidden files/dirctories"))
```
I don't have time to experiment with it right now, so could someone please implement the TS?

**Comment by joehillen** [comment]
@nixmaniack This is awesome. Thank you! <3

**Comment by d12frosted** [comment]
Fixed with released of Spacemacs v0.200. Let us know if you still have any problems with this issue.
